### DISCUSSION TITLE
handle local docs ?

### DISCUSSION THREAD

**Discussion by kalle07** [mudler/LocalAI#2059] [Q&A]
any local docs/pdf  i can talk with ?
some library all indexed read for use   ;)

**Comment by JackBekket** [comment]
I am thinking about same problem

Might be implemented through embeddings (basically to search among documents for relevant parts ("indexing")) and feed chunk of this data as input for model, so we can have RAG

I have planned to test how just embeddings work with local-ai and langchain, so I can search through doc or database, to find what I want using context similarity search 

For example, it might be useful through parsing user's data or to working with historical data

Let's say you have a db with users-locations, which can be any location in any format like "Rim, Italy  -- Rome, Italy -- Berlin -- Bay Area -- НижнийНовгород -- Украина Киев -- Украiна Кiев -- worlidwide -- Moon(planet earh)" and so on
And you can fix this problem using embeddings and vector-store, so if you looking for USA then you can find those who inputed 'NY' and so on


Working with historical data means, that you can just add dump of all local-news for a few years and make your model aware of context nowdays. Or it could be chatHistory with chat-sessions, or it could be CVE list dump, or even your own codebase

but that just embeddings

**Comment by JackBekket** [comment]
But for a full RAG we also need some kind of `retrival` ?

**Comment by JackBekket** [comment]
I found some kind of solution here:
https://github.com/build-on-aws/rag-golang-postgresql-langchain/blob/main/query.go

**Comment by BradKML** [comment]
Seconding this since a lot of agentic libraries have RAGs built in with reranking and embedding functions.

**Comment by kalle07** [comment]
why you thinking of it ?
GPT4ALL and Jan and privateGPT .. all have it
you only must look and copy and paste the best of all  ;)

**Reply by BradKML** [reply]
Could you do a whole list on what is available? Not sure if LMStudio does it but it ain't FOSS

**Comment by kalle07** [comment]
lmstudio work at the moment only via an embedding model and send it via server to eg privateGPT


i know only
GPT4all
https://github.com/nomic-ai/gpt4all

jan
https://github.com/janhq/jan

privategpt
https://github.com/zylon-ai/private-gpt

maybe it give 2more, but if, it was to complicate to install
